pkg/helpers/tokencmd/negotiator_gssapi.go

```diff
@@ -1,197 +1,77 @@
-	flags uint32
-
 	// principalName contains the name of the principal desired by the user, if specified.
 	principalName string
-	// cred holds the credentials obtained for the specified principalName.
-	// if no principalName is specified, GSS_C_NO_CREDENTIAL is used.
-	cred *gssapi.CredId
 
 	// track whether the last response from InitSecContext was GSS_S_COMPLETE
 	complete bool
+
+	client *krb5client.Client
 }
 
 func NewGSSAPINegotiator(principalName string) Negotiator {
 	return &gssapiNegotiator{principalName: principalName}
 }
 
 func (g *gssapiNegotiator) Load() error {
-	_, err := g.loadLib()
-	return err
+	// do nothing
+	return nil
 }
 
+// TODO(lorbus) Make this function more resilient
 func (g *gssapiNegotiator) InitSecContext(requestURL string, challengeToken []byte) (tokenToSend []byte, err error) {
-	lib, err := g.loadLib()
-	if err != nil {
-		return nil, err
-	}
-
-	// Initialize our context if we haven't already
-	if g.ctx == nil {
-
-		if len(g.principalName) > 0 {
-			// Get credentials for a specific principal
-			klog.V(5).Infof("acquiring credentials for principal name %s", g.principalName)
-			credBuffer, err := lib.MakeBufferString(g.principalName)
-			if err != nil {
-				return nil, convertGSSAPIError(err)
-			}
-			defer credBuffer.Release()
-
-			credName, err := credBuffer.Name(lib.GSS_KRB5_NT_PRINCIPAL_NAME)
-			if err != nil {
-				return nil, convertGSSAPIError(err)
-			}
-			defer credName.Release()
-
-			cred, _, _, err := lib.AcquireCred(credName, time.Duration(0), lib.GSS_C_NO_OID_SET, gssapi.GSS_C_INITIATE)
-			if err != nil {
-				klog.V(5).Infof("AcquireCred returned error: %v", err)
-				return nil, convertGSSAPIError(err)
-			}
-			g.cred = cred
-		} else {
-			// otherwise, express no opinion about the credentials and let gssapi decide
-			g.cred = lib.GSS_C_NO_CREDENTIAL
-		}
 
-		serviceName, err := getServiceName('@', requestURL)
+	if g.client == nil {
+		// TODO confirm support for recursive `include` and `includedir` in conf file
+		// TODO Don't hardcode filepath, use env var instead
+		config, err := krb5conf.Load("/etc/krb5.conf")
 		if err != nil {
 			return nil, err
 		}
 
-		klog.V(5).Infof("importing service name %s", serviceName)
-		nameBuf, err := lib.MakeBufferString(serviceName)
+		// TODO add support for using pw instead of kt
+		keytab, err := krb5kt.Load("/etc/krb5.keytab")
 		if err != nil {
-			return nil, convertGSSAPIError(err)
+			return nil, err
 		}
-		defer nameBuf.Release()
 
-		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		// TODO check whether REALM can be safely omitted (2nd param)
+		g.client = krb5client.NewWithKeytab(g.principalName, "", keytab, config)
+
+		err = g.client.Login()
 		if err != nil {
-			return nil, convertGSSAPIError(err)
+			return nil, err
 		}
-		g.name = name
-		g.ctx = lib.GSS_C_NO_CONTEXT
 	}
 
-	incomingTokenBuffer, err := lib.MakeBufferBytes(challengeToken)
+	servicePrincipleName, err := getServiceName('/', requestURL)
```

Comment:
`servicePrincipalName`?

Reply:
Why did you change the principal separator?

Reply:
It's the format the lib expects, see https://github.com/jcmturner/gokrb5/blob/master/v8/USAGE.md#generic-kerberos-client

Reply:
> `servicePrincipalName`?

Yes, I'm reusing the vocabulary that SPNEGO uses here.

Reply:
I meant `s/Principle/Principal` to match the Kerberos terms

Reply:
As for the separator: https://web.mit.edu/kerberos/krb5-1.5/krb5-1.5.4/doc/krb5-user/What-is-a-Kerberos-Principal_003f.html